core/state-machine/src/backend.rs

```diff
@@ -283,29 +357,66 @@ impl<H: Hasher> Clone for InMemory<H> {
 impl<H: Hasher> PartialEq for InMemory<H> {
 	fn eq(&self, other: &Self) -> bool {
 		self.inner.eq(&other.inner)
+			&& self.kv().eq(other.kv()) 
 	}
 }
 
 impl<H: Hasher> InMemory<H> {
+	/// Key value inmemory storage is either used from the
+	/// state db or its own field.
+	/// This function access the right field.
+	fn kv(&self) -> &InMemoryKvBackend {
+		if let Some(kv) = self.kv.as_ref() {
+			kv
+		} else {
+			self.state.as_ref().unwrap().kv_backend()
+		}
+	}
+
+	/// Access the current key value db in a mutable way.
+	fn kv_mut(&mut self) -> &mut InMemoryKvBackend {
+		if let Some(kv) = self.kv.as_mut() {
+			kv
+		} else {
+			self.state.as_mut().unwrap().kv_backend_mut()
+		}
+	}
+
+	/// Extract key values from `InMemory`, by emptying it.
+	fn extract_kv(&mut self) -> InMemoryKvBackend {
+		if let Some(kv) = self.kv.take() {
+			kv
+		} else {
+			std::mem::replace(
+				self.state.as_mut().unwrap().kv_backend_mut(),
+				Default::default(),
+			)	
+		}
+	}
+
 	/// Copy the state, with applied updates
 	pub fn update(&self, changes: <Self as Backend<H>>::Transaction) -> Self {
 		let mut inner: HashMap<_, _> = self.inner.clone();
-		for (storage_key, key, val) in changes {
+		let mut kv: HashMap<_, _> = self.kv().clone();
+		for (storage_key, key, val) in changes.storage {
 			match val {
 				Some(v) => { inner.entry(storage_key).or_default().insert(key, v); },
 				None => { inner.entry(storage_key).or_default().remove(&key); },
 			}
 		}
+		kv.extend(changes.kv);
 
-		inner.into()
+		let kv = Some(kv);
+		InMemory { inner, kv, state: None, _hasher: PhantomData }
 	}
 }
 
 impl<H: Hasher> From<HashMap<Option<Vec<u8>>, HashMap<Vec<u8>, Vec<u8>>>> for InMemory<H> {
 	fn from(inner: HashMap<Option<Vec<u8>>, HashMap<Vec<u8>, Vec<u8>>>) -> Self {
 		InMemory {
 			inner: inner,
-			trie: None,
+			state: None,
+			kv: Some(Default::default()),
```

Comment:
All this from sounds now a bit weird to me because it is not obvious if it should make an InMemory with the kv storage or without the kv storage. usually I prefer a new function that makes it obvious. However I'm ok with this. But if we really want to make kv optional then add a doc in the from implementation stating that kv is activated

Reply:
Yes, it is only to avoid cloning it, would probably be better with a box. But yes the option here does not have meaning, in current state you need either kv being Some or state being Some.
But yes I could probably just clone.